buildSrc/src/main/kotlin/Java9Modularity.kt

```diff
@@ -0,0 +1,63 @@
+import org.gradle.api.Project
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.api.tasks.compile.AbstractCompile
+import org.gradle.api.tasks.compile.JavaCompile
+import org.gradle.kotlin.dsl.register
+import java.util.spi.ToolProvider
+
+object Java9Modularity {
+    @JvmStatic
+    fun configureJava9ModuleInfo(project: Project) {
+        project.afterEvaluate {
+            val jdeps = ToolProvider.findFirst("jdeps")
+            val compileKotlinJvm = tasks.findByName("compileKotlinJvm") as AbstractCompile?
+            val jvmJar = tasks.findByName("jvmJar") as Jar?
+
+            if (compileKotlinJvm != null && jvmJar != null) {
+                val modularityDir = buildDir.resolve("modularity")
+                val generatedDir = modularityDir.resolve("generated")
+                val classesDir = modularityDir.resolve("classes")
+                val jvmGenerateModuleInfo = tasks.register("jvmGenerateModuleInfo", Jar::class) {
+                    dependsOn(compileKotlinJvm)
+                    destinationDirectory.set(modularityDir)
+                    from(jvmJar.source)
+                    exclude { it.file.startsWith(classesDir) }
+                    doLast {
+                        generatedDir.deleteRecursively()
+                        jdeps.orElseThrow { IllegalStateException("Tool 'jdeps' is not available") }.run(
+                            System.out, System.err,
+                            "--multi-release", "9",
+                            "--generate-module-info", generatedDir.toString(),
+                            "--module-path", compileKotlinJvm.classpath.asPath,
+                            archiveFile.get().toString()
+                        )
+                    }
+                }
+                val jvmCompileModuleInfo = tasks.register("jvmCompileModuleInfo", JavaCompile::class) {
+                    dependsOn(jvmGenerateModuleInfo)
+                    destinationDirectory.set(classesDir)
+                    source(generatedDir)
+                    classpath = files()
+                    doFirst {
+                        val moduleName = generatedDir.listFiles().first().name
+                        options.compilerArgs = listOf(
+                            "--release", "9",
+                            "--module-path", compileKotlinJvm.classpath.asPath,
+                            "--patch-module", "$moduleName=${compileKotlinJvm.destinationDir}",
+                            "-Xlint:-requires-transitive-automatic"
```

Comment:
This will fail (as expected) when running build on JDK 8. Do we have a manual about setting up the environment for building kotlinx.serialization? It may need a note about required JDKs and their location.

Reply:
Ah, good point. Is that something I should add to the README?

Reply:
I figured `docs/building.md` would be the correct place.
An alternative would be to only configure the compilation task of the `module-info` if the JDK is version 9 or higher.
Maybe with a warning that the resulting JAR is not JPMS compatible due to the used JDK?